File: app/src/main/java/com/chichiangho/base/widgets/RefreshRecyclerView.java
package com.chichiangho.base.widgets;

import android.content.Context;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.chichiangho.base.utils.DeviceUtil;
import com.chichiangho.base.utils.Logger;
import com.chichiangho.base.R;

public class RefreshRecyclerView extends SwipeRefreshLayout {
    private static final String TAG = "RefreshRecyclerView";
    private static final int REFRESH_INTERVAL = 1000;
    private RefreshListener listener;
    private RecyclerView recyclerView;
    private RelativeLayout footerView;
    private long lastRequestTime;
    private boolean touchEnd;
    private boolean loadingMore;
    private Type managerType;
    private RecyclerView.LayoutManager manager;
    private TextView footTv;
    private ProgressBar footProgress;
    private int hideCount = 1;

    private enum Type {Linear, Grid, StaggeredGrid}

    public RefreshRecyclerView(Context context) {
        super(context);
    }

    public RefreshRecyclerView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        init();
    }

    public void setHideFooterWhenItemLessThan(int itemCount) {
        hideCount = itemCount;
    }

    void init() {
        setColorSchemeResources(android.R.color.holo_blue_light,
                android.R.color.holo_red_light, android.R.color.holo_orange_light,
                android.R.color.holo_green_light);

        recyclerView = (RecyclerView) getChildAt(1);
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            private int lastVisibleItem;

            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);

                if (listener != null && !touchEnd && !isRefreshing() && !loadingMore
                        && recyclerView.getAdapter().getItemCount() > (recyclerView.getAdapter() instanceof RecyclerViewFooterAdapter ? 1 : 0)
                        && newState == RecyclerView.SCROLL_STATE_IDLE && lastVisibleItem + 1 == recyclerView.getAdapter().getItemCount()) {
                    long cur = System.currentTimeMillis();
                    if (cur - lastRequestTime > REFRESH_INTERVAL) {
                        if (recyclerView.getAdapter() instanceof RecyclerViewFooterAdapter) {
                            initFooter();
                            if (recyclerView.getAdapter().getItemCount() > hideCount) {
                                footerView.setVisibility(VISIBLE);
                                footTv.setText("loading...");
                                footProgress.setVisibility(VISIBLE);
                            } else {
                                footerView.setVisibility(GONE);
                            }
                        }
                        lastRequestTime = cur;
                        loadingMore = true;
                        listener.onLoadMore();
                    }
                }
            }

            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (manager == null) {
                    manager = recyclerView.getLayoutManager();
                    if (manager instanceof LinearLayoutManager)
                        managerType = Type.Linear;
                    else if (manager instanceof GridLayoutManager)
                        managerType = Type.Grid;
                    else if (manager instanceof StaggeredGridLayoutManager)
                        managerType = Type.StaggeredGrid;
                }
                switch (managerType) {
                    case Linear:
                        lastVisibleItem = ((LinearLayoutManager) manager).findLastVisibleItemPosition();
                        break;
                    case Grid:
                        lastVisibleItem = ((GridLayoutManager) manager).findLastVisibleItemPosition();
                        break;
                    case StaggeredGrid:
                        int[] positions = new int[((StaggeredGridLayoutManager) manager).getSpanCount()];
                        ((StaggeredGridLayoutManager) manager).findLastVisibleItemPositions(positions);
                        int max = 0;
                        for (int value : positions) {
                            if (value > max) {
                                max = value;
                            }
                        }
                        lastVisibleItem = max;
                        break;
                    default:
                        Logger.e(TAG, "LayoutManager must be LinearLayoutManager or GridLayoutManager or StaggeredGridLayoutManager");
                        break;
                }
            }
        });

        setOnRefreshListener(new OnRefreshListener() {
            @Override
            public void onRefresh() {
                long cur = System.currentTimeMillis();
                if (listener != null && !loadingMore && cur - lastRequestTime > REFRESH_INTERVAL) {
                    touchEnd = false;
                    listener.onRefresh();
                } else {
                    setRefreshing(false);
                }
            }
        });
    }

    private void initFooter() {
        if (footerView == null) {
            footerView = (RelativeLayout) LayoutInflater.from(getContext()).inflate(R.layout.layout_refresh_footer, null, false);
            RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(DeviceUtil.getScreenWidth(), (int) getResources().getDimension(R.dimen.refresh_footer_height));
            footerView.setLayoutParams(params);
            footTv = (TextView) footerView.findViewById(R.id.footer_tv);
            footProgress = (ProgressBar) footerView.findViewById(R.id.footer_progress);

            ((RecyclerViewFooterAdapter) recyclerView.getAdapter()).setFooterView(footerView);
        }
    }

    @Override
    public void setRefreshing(final boolean refreshing) {
        loadingMore = false;
        if (recyclerView.getAdapter() instanceof RecyclerViewFooterAdapter) {
            initFooter();
            if (recyclerView.getAdapter().getItemCount() > hideCount) {
                footerView.setVisibility(VISIBLE);
                if (!touchEnd)
                    footTv.setText("上拉加载更多");
                footProgress.setVisibility(GONE);
            } else {
                footerView.setVisibility(GONE);
            }
        }
        super.setRefreshing(refreshing);
    }

    public void setNoMoreData() {
        touchEnd = true;
        if (footerView != null) {
            if (recyclerView.getAdapter().getItemCount() > hideCount) {
                footerView.setVisibility(VISIBLE);
                footTv.setText("我是有底线的");
                footProgress.setVisibility(GONE);
            } else {
                footerView.setVisibility(GONE);
            }
        }
    }

    public void setListener(RefreshListener listener) {
        this.listener = listener;
    }

    public interface RefreshListener {
        void onRefresh();

        void onLoadMore();
    }
}

File: app/src/main/java/com/chichiangho/base/widgets/RefreshRecyclerView.kt
package com.chichiangho.base.widgets

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.widget.ProgressBar
import android.widget.RelativeLayout
import android.widget.TextView

import com.chichiangho.base.R

class RefreshRecyclerView : SwipeRefreshLayout {
    private var listener: RefreshListener? = null
    private var recyclerView: RecyclerView? = null
    private var touchEnd: Boolean = false//是否有更多数据，true没有
    private var notFill: Boolean = false//数据是否充满RecyclerView，true没有
    private var loadingMore: Boolean = false
    private var pulling: Boolean = false//是否在上拉手势中，true在
    private var footerView: RelativeLayout? = null
    private var footTv: TextView? = null
    private var footProgress: ProgressBar? = null
    private var valueAnimator: ValueAnimator? = null


    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    override fun onFinishInflate() {
        super.onFinishInflate()
        init()
    }

    internal fun init() {
        setColorSchemeResources(android.R.color.holo_blue_light,
                android.R.color.holo_red_light, android.R.color.holo_orange_light,
                android.R.color.holo_green_light)

        for (i in 0 until childCount) {
            if (getChildAt(i) is RecyclerView) {
                recyclerView = getChildAt(i) as RecyclerView
                break
            }
        }
        recyclerView ?: throw Exception("this view should got a recyclerView as child")

        recyclerView?.addItemDecoration(object : RecyclerView.ItemDecoration() {
            override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State?) {
                super.getItemOffsets(outRect, view, parent, state)
                outRect.bottom = 1
            }
        })//由于某些原因，无ItemDecoration时canScrollVertically()函数无法正确判断，故添加一个1px的ItemDecoration以修复
        recyclerView?.setOnTouchListener(object : View.OnTouchListener {
            var distance: Float = 0f
            var mFirstY: Float = 0f
            var motionStart: Float = 0f

            override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
                initFooter()
                if (touchEnd || notFill || loadingMore || isRefreshing)
                    return false

                if (valueAnimator?.isRunning == true)
                    return false

                when (motionEvent.action) {
                    MotionEvent.ACTION_MOVE -> {
                        if (motionStart == 0f)
                            motionStart = motionEvent.rawY
                        if (!pulling && recyclerView?.canScrollVertically(1) != true) {
                            mFirstY = motionEvent.rawY
                            pulling = true
                        }
                        if (pulling) {
                            distance = (mFirstY - motionEvent.rawY) * 6 / 10//0.6的ratio
                            val params = footerView?.layoutParams
                            params?.height = distance.toInt()
                            footerView?.layoutParams = params
                        }
                    }
                    MotionEvent.ACTION_UP -> {
                        pulling = false
                        val motionEnd = motionEvent.rawY
                        if (distance > 0 && motionStart + 50 > motionEnd) {
                            motionStart = 0f
                            if (footerView != null) {
                                valueAnimator = ObjectAnimator.ofFloat(distance, resources.getDimension(R.dimen.refresh_footer_height)).setDuration(ANIME_TIME)
                                valueAnimator?.addUpdateListener { animation ->
                                    val nowDistance = animation.animatedValue as Float
                                    val layoutParams = footerView?.layoutParams
                                    layoutParams?.height = nowDistance.toInt()
                                    footerView?.layoutParams = layoutParams
                                }
                                valueAnimator?.addListener(object : AnimatorListenerAdapter() {
                                    override fun onAnimationEnd(animation: Animator) {
                                        super.onAnimationEnd(animation)
                                        footTv?.text = "加载中..."
                                        footProgress?.visibility = View.VISIBLE

                                        if (listener != null) {
                                            loadingMore = true
                                            loadMore()
                                        }
                                    }
                                })
                                valueAnimator?.start()
                            } else {
                                if (listener != null) {
                                    loadingMore = true
                                    loadMore()
                                }
                            }
                        }
                    }
                }
                return false
            }
        })

        setOnRefreshListener {
            if (listener != null && !pulling && !loadingMore) {
                if (footerView != null) {
                    val layoutParams = footerView?.layoutParams
                    layoutParams?.height = 0
                    footerView?.layoutParams = layoutParams
                }
                touchEnd = false
                refresh()
            } else {
                isRefreshing = false
            }
        }
    }

    private fun loadMore() {
        postDelayed({ listener?.onLoadMore() }, ANIME_TIME)
    }

    private fun refresh() {
        postDelayed({ listener?.onRefresh() }, ANIME_TIME)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        notFill = !(recyclerView?.canScrollVertically(1) ?: false || recyclerView?.canScrollVertically(-1) ?: false)
    }

    private fun initFooter() {
        if (footerView == null && recyclerView?.adapter is RecyclerViewFooterAdapter<*>) {
            footerView = LayoutInflater.from(context).inflate(R.layout.layout_refresh_footer, null, false) as RelativeLayout
            footTv = footerView?.findViewById(R.id.footerTv)
            footProgress = footerView?.findViewById(R.id.footerProgress)
            val params = RelativeLayout.LayoutParams(measuredWidth, 0)
            footerView?.layoutParams = params
            footTv?.text = "上拉加载更多"
            footProgress?.visibility = View.GONE
            (recyclerView?.adapter as RecyclerViewFooterAdapter<*>).footerView = footerView
        }
    }

    override fun setRefreshing(refreshing: Boolean) {
        if (footerView != null) {
            if (loadingMore) {
                val start: Float = footerView?.layoutParams?.height?.toFloat() as Float
                valueAnimator = ObjectAnimator.ofFloat(start, 0f).setDuration(ANIME_TIME)
                valueAnimator?.addUpdateListener { animation ->
                    val nowDistance = animation.animatedValue as Float
                    val params = footerView?.layoutParams
                    params?.height = nowDistance.toInt()
                    footerView?.layoutParams = params
                }
                valueAnimator?.addListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator) {
                        super.onAnimationEnd(animation)
                        if (touchEnd) {
                            footerView?.layoutParams?.height = resources.getDimension(R.dimen.refresh_footer_height).toInt()
                            footTv?.text = "我是有底线的"
                            footProgress?.visibility = View.GONE
                        } else {
                            footTv?.text = "上拉加载更多"
                            footProgress?.visibility = View.GONE
                        }
                        loadingMore = false
                    }
                })
                valueAnimator?.start()
            } else {
                footTv?.text = "上拉加载更多"
                footProgress?.visibility = View.GONE
            }
        } else {
            loadingMore = false
        }

        super.setRefreshing(refreshing)
    }

    fun setNoMoreData() {
        touchEnd = true
        if (valueAnimator?.isRunning == false) {
            footerView?.layoutParams?.height = resources.getDimension(R.dimen.refresh_footer_height).toInt()
            footTv?.text = "我是有底线的"
            footProgress?.visibility = View.GONE
        }
    }

    fun setListener(listener: RefreshListener) {
        this.listener = listener
    }

    interface RefreshListener {
        fun onRefresh()

        fun onLoadMore()
    }

    companion object {
        private val TAG = "RefreshRecyclerView"
        private val ANIME_TIME = 400L
    }
}